securedrop_client/gui/widgets.py

```diff
@@ -849,20 +849,37 @@ def update(self, sources: List[Source]):
         current_source_id = current_source and current_source.id
 
         self.clear()
-
-        for source in sources:
-            new_source = SourceWidget(source)
-            new_source.setup(self.controller)
-            self.source_widgets[source.uuid] = new_source
-
-            list_item = QListWidgetItem(self)
-            list_item.setSizeHint(new_source.sizeHint())
-
-            self.addItem(list_item)
-            self.setItemWidget(list_item, new_source)
-
-            if source.id == current_source_id and has_source_selected:
-                self.setCurrentItem(list_item)
+        sources.reverse()
+        self.add_source(sources, current_source_id, has_source_selected)
+
+    def add_source(self, sources, current_source_id, has_source_selected, slice_size=1):
+        """
+        Add a slice of sources, ensure the currently selected source is
+        retained and, if necessary, reschedule the addition of more sources.
+        """
+        def schedule_source_creation(slice_size=slice_size):
+            if sources:
```

Comment:
looks like the code in the closure is under `if sources` to terminate recursion? if so we could do at the top of the closure something like: 

```py
if not sources:  # Finished drawing all sources
    return
```

or if we want to prevent the final call of `self.add_source` in the base case we could do:

```py
remaining_sources = sources[slice_size:]
if remaining_sources:
    self.add_source(remaining_sources, current_source_id, has_source_selected, new_slice_size)
```

Reply:
Correct. I did have a check at the beginning of the of the closure, but terminated recursion via the conditional instead for what felt like a good reason at the time, but for the life of me, I can't remember why. Happy to refactor for the sake of readability.

Reply:
gotcha, that makes sense. yeah let's make a slight refactor to do one of these two options (either works) for readability (reduces indentation level which makes things easier to follow for me at least 😇)